mobilecoind/src/sync.rs

```diff
@@ -286,8 +296,115 @@ fn sync_thread_entry_point(
     }
 }
 
-/// Sync a single monitor.
-fn sync_monitor(
+/// Sync a single monitor, spawning parallel jobs to perform block scanning if
+/// it is many blocks behind. This allows us to use all the CPU cores scanning
+/// even when we only have a single monitor to work on, and all the blocks are
+/// pretty small (only 2 or 3 UTXOs), which is a very common configuration and
+/// situation. First, we test how far behind the monitor is. When the monitor is
+/// close to caught up, we fallback to sync_monitor_sequential.
+fn sync_monitor_parallel(
+    ledger_db: &LedgerDB,
+    mobilecoind_db: &Database,
+    monitor_id: &MonitorId,
+    num_workers: usize,
+    logger: &Logger,
+) -> Result<SyncMonitorOk, Error> {
+    let monitor_data = mobilecoind_db.get_monitor_data(monitor_id)?;
+    let num_blocks = ledger_db.num_blocks()?;
+
+    // If we cannot hope to work on num_workers blocks in parallel for this montior,
+    // then we are pretty close to the end, let's fall back to
+    // sync_monitor_sequential. Otherwise, we should be able to get num_workers
+    // blocks and work on them in parallel.
+    if monitor_data.next_block + num_workers as u64 >= num_blocks {
+        return sync_monitor_sequential(ledger_db, mobilecoind_db, monitor_id, logger);
+    }
+
+    // Each worker will try to call
+    // ledger_db.get_block_contents(monitor_data.next_block + worker_idx)
+    // and then
+    // match_tx_outs_into_utxos
+    // in parallel.
+    //
+    // At the end we collect the results and add them to the database in order,
+    // because the database needs that to be done sequentially. But the scanning
+    // is the slow part here, so this is fine.
+    let parallel_results = (0..num_workers)
+        .into_par_iter()
+        .map(
+            |worker_idx| -> Result<(Vec<UnspentTxOut>, Vec<KeyImage>), Error> {
+                // We don't expect to get Error::NotFound here, since we earlier tested
+                // ledger_db.num_blocks(). The other error cases are not common
+                // either.
+                let block_idx = monitor_data.next_block + worker_idx as u64;
+                let block_contents = match ledger_db.get_block_contents(block_idx) {
+                    Ok(block_contents) => block_contents,
+                    // Note: mc_ledger_db::Error::NotFound is being handled like any other error
+                    // here, since it really isn't expected.
+                    Err(err) => {
+                        return Err(err.into());
```

Comment:
So cant this whole `match` block get replaced with `let block_contents = match ledger_db.get_block_contents(block_idx)?;`?

Reply:
yeah you are right I guess